### ISSUE TITLE
The unstable state of the wazuh-states-vulnerabilities-wazuh index

### ISSUE THREAD

**Issue by blurryface-9** [wazuh/wazuh#26118]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.8.0-1 | vulnerability-detection | Cluster | Packages | Ubuntu 22.04 |

After updating Wazuh from version 4.6 to 4.8, we encountered a problem in detecting vulnerabilities on agents. Immediately after the update, a huge number of vulnerabilities were added to the index (significantly more than were discovered before the update). As expected, no alerts were generated for these detections (creating a basic profile for the agent). However, after the first scan, the number of vulnerabilities in the index is constantly changing, while alerts are created for a very small number of vulnerabilities.

During the investigation of this problem, we found that after changing the offset of CTI, a mass scan of agents is launched and for some reason some vulnerabilities are added as new (the field 'vulnerability.detected_at' changes to the current date), some vulnerabilities remain with the same detection date. Alerts for these events are also not created.

For example:
```
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] packageScanner.hpp:477 at versionMatch(): DEBUG: Match found, the package 'python2.7-minimal', is vulnerable to 'CVE-2022-0391'. Current version: '2.7.18-1~20.04.3' (less than '2.7.18-1~20.04.3+esm1' or equal to ''). - Agent 'xxx' (ID: '111', Version: 'v4.8.0').
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] packageScanner.hpp:477 at versionMatch(): DEBUG: Match found, the package 'python2.7-minimal', is vulnerable to 'CVE-2022-0391'. Current version: '2.7.18-1~20.04.3' (less than '2.7.18-1~20.04.3+esm1' or equal to ''). - Agent 'xxx' (ID: '111', Version: 'v4.8.0').
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] packageScanner.hpp:477 at versionMatch(): DEBUG: Match found, the package 'python2.7-minimal', is vulnerable to 'CVE-2023-27043'. Current version: '2.7.18-1~20.04.3' (less than '2.7.18-1~20.04.4+esm3' or equal to ''). - Agent 'xxx' (ID: '111', Version: 'v4.8.0').
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] packageScanner.hpp:477 at versionMatch(): DEBUG: Match found, the package 'python2.7-minimal', is vulnerable to 'CVE-2024-7592'. Current version: '2.7.18-1~20.04.3' (less than '2.7.18-1~20.04.4+esm2' or equal to ''). - Agent 'xxx' (ID: '111', Version: 'v4.8.0').
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] eventInsertInventory.hpp:115 at handleRequest(): DEBUG: Inserting agent package key: 111_9ab6e60df8b4ag90f6009oiuf081ddbcea44e468 -> CVE-2023-27043,CVE-2024-7592,CVE-2021-4189,CVE-2022-0391
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] eventDetailsBuilder.hpp:101 at handleRequest(): DEBUG: Building event details for component type: 1
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] resultIndexer.hpp:62 at handleRequest(): DEBUG: Processing and publish key: CVE-2023-27043
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] resultIndexer.hpp:62 at handleRequest(): DEBUG: Processing and publish key: CVE-2024-7592
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] resultIndexer.hpp:62 at handleRequest(): DEBUG: Processing and publish key: CVE-2021-4189
2024/10/03 01:38:02 wazuh-modulesd:vulnerability-scanner[3858723] resultIndexer.hpp:62 at handleRequest(): DEBUG: Processing and publish key: CVE-2022-0391
```

4 vulnerabilities were found in the python2.7-minimal package after scanning on 03 Oct. They are all added to the index, but the vulnerability.detected_at field changed to 3 Oct only for CVE-2023-27043. For the rest of the vulnerabilities, the previous value of the field has been preserved (which in fact also does not reflect the actual date of detection).

![image](https://github.com/user-attachments/assets/8e6acadb-4516-48f7-a2f6-2e0d1a69d16e)

No alerts were generated:

![image](https://github.com/user-attachments/assets/0107aa88-b2a4-41b3-bcb3-1a5bac2a3f96)

At the same time, no actions are performed with the package, and most likely these vulnerabilities were there even before the update.

When testing a vulnerability-detection  module with the addition/removal of a vulnerable package, alerts are generated as expected.

Thus, the main problem is the unstable number of vulnerabilities in the index without alerts.

**Comment by pereyra-m** [comment]
Hello @blurryface-9 

I'll try to give you more details about the module behavior to make the situation clearer.

It's expected that not all changes of the vulnerabilities inventory trigger alerts. We will give more details about this here
- https://github.com/wazuh/wazuh-documentation/issues/7832

This is done on purpose to avoid the generation of alerts that aren't triggered by "real time" events, for example:
- A new offset added to the content will trigger the re-scan of all agents, all the solved vulnerabilities will be removed from the inventory and the new ones will be added; but no alerts will be generated in this situation.
- An agent that moves from one node to another and re-syncs its inventory. The worker node will receive and scan the packages that it didn't have in its inventory but won't trigger alerts.

> When testing a vulnerability-detection module with the addition/removal of a vulnerable package, alerts are generated as expected.

This situation is covered in the normal **syscollector** scan, that is the only situation that will generate alerts.

>  for some reason some vulnerabilities are added as new (the field 'vulnerability.detected_at' changes to the current date), some vulnerabilities remain with the same detection date. 

Another characteristic of the module is that the vulnerabilities index won't update the entries that it already has, only adds or removes elements.
So if, for example, your agent has some specific OS and Wazuh agent version and some vulnerabilities are indexed with this information, a further upgrade of the OS/agent will index any new vulnerability with the current information but the older record **won't** be updated. This only will change during a re-scan, that removes all old data from the index and inserts the last scan results from scratch.

I can't reproduce in my local environment the situation you've shown, where the detection date isn't the same for all the CVEs results of the same scan. But I'll talk with @wazuh/devel-cppserver-div1 if this field is behaving as expected.

**Comment by blurryface-9** [comment]
Hello @pereyra-m, thank you for your reply!
I would like to make it clear that in our scheme, there are load balancers between the agents and the nodes of the manager (one per manager node). Agents use a single domain name to connect to managers, behind which all load balancers are located. Routing is performed via BGP. Therefore, agents can periodically change the manager node to which they are connected. Could this be the cause of our index instability problem? Should we come up with a way to strictly bind agents to manager nodes?
At the moment, we have switched all agents directly to the nodes of the manager (bypassing the load balancers) and continue to monitor the status of the index. 

> A new offset added to the content will trigger the re-scan of all agents, all the solved vulnerabilities will be removed from the inventory and the new ones will be added; but no alerts will be generated in this situation.

I would also like to clarify this point. For example, we have a package on the agent for which there are no vulnerabilities in CTI. At some point, a vulnerability for this package is discovered and gets into CTI. The offset is updated, and the agent's rescan begins. This vulnerability is indexed, but no alert is generated, right? Wouldn't an alert in such a case be useful?

**Comment by pereyra-m** [comment]
Hi again @blurryface-9 

Now I better understand your use case and environment setup.
Can you tell me if the agents update their packages' inventory frequently?

> Therefore, agents can periodically change the manager node to which they are connected. Could this be the cause of our index instability problem?

The vulnerability scanner must work in both single node and cluster configurations, but the latter presents some complexities due to the Wazuh current architecture.

The worker nodes are meant to share the workload of processing events, and they share the agents' keys and some basic information with the master. But the packages inventory isn't synced across the cluster, and each node will have a local packages snapshot of the agents in the state it was the last time the agent was connected to them.
Also, these nodes have a local DB with the vulnerabilities associated with each agent and this isn't synced either.

Considering that we have only one unified index for all the environment, the nodes don't (and shouldn't) sync at the same time their vulnerabilities inventories because the data isn't consistent. To overcome this problem, only the node the agent is currently connected to will sync the index with its local vulnerabilities index after a **syscollector** scan (1 hour by default). 

So you can imagine that an agent moving from one node to another forces:
- Package inventory pairing
- Vulnerabilities inventory pairing
- Index sync

This could give the temporary impression that the index is constantly changing. But if the agents' inventory doesn't change too often, after a while no matter which node the agent is talking to, all of them will have the last and valid packages inventory and will sync the same data.

Nevertheless, if this isn't the case here, please share more details using one agent as an example. There could be a hidden bug.

> For example, we have a package on the agent for which there are no vulnerabilities in CTI. At some point, a vulnerability for this package is discovered and gets into CTI. The offset is updated, and the agent's rescan begins. This vulnerability is indexed, but no alert is generated, right?

This sequence is correct and currently, the module won't generate alerts about this. I'll share your comments with the team to analyze this requirement. But there isn't a configuration to change this behavior.

**Comment by blurryface-9** [comment]
Hi again @pereyra-m! 
Thank you so much for your help. Sorry for taking so long to respond. I had to spend some time analyzing our situation.

> Can you tell me if the agents update their packages' inventory frequently?

The package inventory is rarely changed on agents in our infrastructure.

After experimenting with connecting agents directly to the manager's nodes, at first sight we succeeded in stabilizing the state of vulnerabilities in the index. 

At the same time, we also updated the agents from version 4.6.0 to 4.8.0 and found that the number of vulnerabilities on the agents had changed. For example, an agent that was connected directly to one particular node of the manager had a certain number of vulnerabilities and their number did not change during the observation period (~3-4 days). Then we updated the agent and the rescan started. There were no other changes to the host packages. At the same time, we see that there are more vulnerabilities in the index. Alerts were also not created. And we found similar behavior on all the agents observed in the experiment. Maybe you can guess what this is related to?

We also encountered a problem in which, after switching agents directly to nodes, all vulnerabilities related to agents connected to one particular node were removed from the index. After viewing the logs, we found out that when switching agents got to a node that they had not connected to before and all documents on these agents were deleted from the index:
```
indexer-connector[1703917] indexerConnector.cpp:267 at operator()(): DEBUG: Response: {"took":1254,"errors":false,"items":[{"delete":{"_index":"wazuh-states-vulnerabilities-wazuh","_id":"067_00622e14dc2f0729ce4c5a218e6e46bc86c20020_CVE-xxx-xxx","_version":2,"result":"deleted","_shards":{"total":1,"successful":1,"failed":0},"_seq_no":2378805,"_primary_term":22,"status":200}},...
```
Then, after the rescan, the manager was unable to upload the vulnerabilities found for these agents to the index with errors:
```
wazuh-modulesd:vulnerability-scanner[2200771] scanOrchestrator.hpp:153 at operator()(): ERROR: Error processing delayed event: Couldn't find column family: 'os_initial_scan'
wazuh-modulesd:vulnerability-scanner[1703917] scanAgentList.hpp:253 at handleRequest(): ERROR: Error handling request: Error putting data: Invalid argument: Invalid column family specified in write batch
wazuh-modulesd:vulnerability-scanner[1703917] scanOrchestrator.hpp:153 at operator()(): ERROR: Error processing delayed event: Error deleting data: Invalid argument: Invalid column family specified in write batch
```

According to the log, we have already had a similar problem after the update. Restarting the manager helped solve this problem, but it is unclear what this could be related to, since we do not affect the mapping in the indexer, and whether it will happen again.

**Comment by pereyra-m** [comment]
Hi @blurryface-9 

>  Sorry for taking so long to respond. I had to spend some time analyzing our situation.

Don't worry. In fact, your detailed analysis and report help to improve the module.
These are interesting tests in a cluster configuration.

> The package inventory is rarely changed on agents in our infrastructure.

Thank you for this clarification. 

I'll try to answer each topic mentioned.

> After experimenting with connecting agents directly to the manager's nodes, at first sight we succeeded in stabilizing the state of vulnerabilities in the index.

In this simple setup, if all agents talk to the manager then it's reasonable the index is more stable.

> At the same time, we also updated the agents from version 4.6.0 to 4.8.0 and found that the number of vulnerabilities on the agents had changed.

This is interesting. It'd be really valuable if you could share the OS details of those agents, and the example packages inventory of one of them. The **syscollector** capability has received some updates in v4.7.0 and v4.8.0 that might explain this

https://github.com/wazuh/wazuh/releases/tag/v4.7.0
- Added package inventory support for MacPorts in Syscollector. (https://github.com/wazuh/wazuh/pull/15877)
- Added package inventory support for PYPI and node in Syscollector. (https://github.com/wazuh/wazuh/pull/17982)

https://github.com/wazuh/wazuh/releases/tag/v4.8.0
- Added snap package manager support to Syscollector. (https://github.com/wazuh/wazuh/pull/15740)

Can you confirm that after the upgrade, you now have more packages detected for those agents?
Considering that the re-scan of the agents triggered after new content is available is the same as the re-scan after an agent upgrade, the detection of new packages seems to be the most probable explanation.

Also, remember that the content is updated often. Maybe new offsets were published in the middle of the upgrade, and the re-scan after the upgrade wasn't run with the same content as before, resulting in more CVEs detected.

> We also encountered a problem in which, after switching agents directly to nodes, all vulnerabilities related to agents connected to one particular node were removed from the index. After viewing the logs, we found out that when switching agents got to a node that they had not connected to before and all documents on these agents were deleted from the index:

As described before, each node has a different list of vulnerabilities for each agent and only the node the agent is connected to will sync its local DB with the index.
In this case, if the agent hasn't connected to that node before, the list is empty and it makes sense that all the vulnerabilities are removed. But shortly, the agent should report all the packages, the node should detect the vulnerabilities, update its local DB, and index all the vulnerabilities again.
Later, the next **syscollector** scan will trigger the re-indexation (there is a 30min abuse control feature that prevents the sync from happening too often) to make sure everything is up to date.

I've run a local test of this situation, and with debug logs enabled, I can see the vulnerabilities are first removed

```
2024/10/17 20:14:42 indexer-connector[2819] indexerConnector.cpp:437 at operator()(): DEBUG: Syncing agent '003' with the indexer.

2024/10/17 20:14:43 indexer-connector[2819] indexerConnector.cpp:267 at operator()(): DEBUG: Response: {"took":549,"errors":false,"items":[{"delete":{"_index":"wazuh-states-vulnerabilities-wazuh","_id":"003_022d75f04e03b48f59841a7a2cbabea0d6d32210_CVE-2024-6119","_version":2,"result":"deleted","_shards":{"total":1,"successful":1,"failed":0} ...
```

But then indexed after the scan ends

```
2024/10/17 20:15:03 indexer-connector[2819] indexerConnector.cpp:414 at operator()(): DEBUG: Response: {"took":991,"errors":false,"items":[{"index":{"_index":"wazuh-states-vulnerabilities-wazuh","_id":"003_dbdeef188ca48c1cff04b34de8a5c682d2e762d9_CVE-2024-45492","_version":3,"result":"created","_shards":{"total":1,"successful":1,"failed":0}, ....
```

> Then, after the rescan, the manager was unable to upload the vulnerabilities found for these agents to the index with errors:

In your environment, maybe that error prevented the proper re-indexation of the vulnerabilities.
That is a race condition we've found and it's being solved here
- https://github.com/wazuh/wazuh/issues/26212

The next content update will trigger a full re-scan of all the agents, so the CVEs should be indexed.

**Comment by blurryface-9** [comment]
Hello @pereyra-m 
Thank you for your clarifications!

> It'd be really valuable if you could share the OS details of those agents, and the example packages inventory of one of them.

Agent's OS is Ubuntu 20.04.5 LTS (Focal). The agent's inventory - [packages.csv](https://github.com/user-attachments/files/17459140/packages.csv)

We also have a question about vulnerability detection again. Recently, a huge number of vulnerabilities (~3000 per agent) were added to the index in a short period of time. The vast majority of entries relate to the linux-image package of different versions. There were no changes related to this package on the agents. 
According to the logs, we see that the offset was updated at this time. It seems strange that vulnerabilities are being discovered even in 2013, are they really only now in the feeds? 
Can you clarify, is this the normal behavior of the module? It is also confusing that such a large number are detected at a time.

> Can you confirm that after the upgrade, you now have more packages detected for those agents?

Unfortunately, the agent's inventory does not provide information about the date when the package was added to the system, so I cannot say for sure if there are more packages.

> Also, remember that the content is updated often. Maybe new offsets were published in the middle of the upgrade, and the re-scan after the upgrade wasn't run with the same content as before, resulting in more CVEs detected.

According to the logs, the offset did not change during the update of the agents.

**Comment by pereyra-m** [comment]
Hi @blurryface-9 !

Thank you for the information provided.

> Agent's OS is Ubuntu 20.04.5 LTS (Focal). The agent's inventory - [packages.csv](https://github.com/user-attachments/files/17459140/packages.csv)

> Unfortunately, the agent's inventory does not provide information about the date when the package was added to the system, so I cannot say for sure if there are more packages.

> According to the logs, the offset did not change during the update of the agents.

Now that I can analyze the packages of the agent and I know its OS, I can confirm that the new features of the 4.7.x and 4.8.x have detected new Python packages that surely weren't available before. This explains the possibility of new vulnerabilities after the agent update because the content wasn't updated during of the upgrade. Still, if you need a deeper analysis, the CVEs and packages detected before and after are required to compare them carefully.

> We also have a question about vulnerability detection again. Recently, a huge number of vulnerabilities (~3000 per agent) were added to the index in a short period of time. The vast majority of entries relate to the linux-image package of different versions. There were no changes related to this package on the agents.
According to the logs, we see that the offset was updated at this time. It seems strange that vulnerabilities are being discovered even in 2013, are they really only now in the feeds?
Can you clarify, is this the normal behavior of the module? It is also confusing that such a large number are detected at a time.

This is possible, and can happen after a change in the content.
Here another user has reported that the package `linux-image` vulnerabilities weren't being reported. 
- https://github.com/wazuh/wazuh/issues/26292

The team research confirmed that there was a package name mismatch preventing the scanning so a translation was required. Now that the change has been uploaded, all Wazuh users can benefit from the fix and the package is being correctly detected.

**Comment by pereyra-m** [comment]
You're welcome @blurryface-9, glad to know that this helps you.

It's a different topic, but I can give a short explanation.
The alert you've shown isn't related to the vulnerability detector, it comes from the **logcollector** module after reading a system file that logs the activity related to the package manager. The analysis for that alert does deserve another issue, where we could verify if it was generated for all the agents, etc.

On the other side, the alerts from the vulnerability detector module won't be always generated, it depends on many factors.
Recently, the documentation was updated to give more details about this recurrent question because we've noticed that it wasn't clear to the users

[Alert generation
](https://documentation.wazuh.com/current/user-manual/capabilities/vulnerability-detection/how-it-works.html#alert-generation)

> Considering all possible situations, the most reliable way to determine the current vulnerabilities in the environment is to query the vulnerability inventory.

By design, the alerts might not be generated for every change in the inventory and that depends on the nature of the triggering event. This behavior can't be configured.
So the current status of the agents will always be the vulnerabilities inventory, even when in the past the alerts were much more frequent.
